### PR TITLE
Simplify C++ ValueFactoryManager

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR simplifies the C++ ValueFactoryManager, renames ValueFactoryFunc to simplify ValueFactory, and updates/simplifies Python/PHP/Ruby use of ValueFactoryManager.

### CODE REVIEW

File: python/modules/IcePy/ValueFactoryManager.h [hunk 1]
```diff
@@ -16,51 +16,46 @@ namespace IcePy
 
     bool initValueFactoryManager(PyObject*);
 
-    class FactoryWrapper final : public Ice::ValueFactory
+    // The IcePy C++ value factory abstract base class.
```
Comment (bernardnormier) [inline python/modules/IcePy/ValueFactoryManager.h:19]: Python and Ruby are almost identical. PHP is a little different, not sure why.

File: python/modules/IcePy/ValueFactoryManager.h [hunk 2]
```diff
@@ -16,51 +16,46 @@ namespace IcePy
...
+    using CustomValueFactoryPtr = std::shared_ptr<CustomValueFactory>;
 
-    class DefaultValueFactory final : public Ice::ValueFactory
+    // The default Python value factory as a C++ value factory.
+    class DefaultValueFactory final : public ValueFactory
     {
     public:
         std::shared_ptr<Ice::Value> create(std::string_view) final;
-
-        void setDelegate(const Ice::ValueFactoryPtr&);
```
Comment (bernardnormier) [inline python/modules/IcePy/ValueFactoryManager.h:41]: The whole delegate thing did not make sense to me, so I removed it.

If the user registers a value factory with type ID "", we can just store it in the C++ map with this string, like any other type ID.

File: python/modules/IcePy/ValueFactoryManager.h [hunk 3]
```diff
@@ -16,51 +16,46 @@ namespace IcePy
...
-    class FactoryWrapper final : public Ice::ValueFactory
+    // The IcePy C++ value factory abstract base class.
+    struct ValueFactory
+    {
+        virtual std::shared_ptr<Ice::Value> create(std::string_view) = 0;
+    };
+    using ValueFactoryPtr = std::shared_ptr<ValueFactory>;
+
+    // Adapts a Python value factory to our C++ value factory.
+    class CustomValueFactory final : public ValueFactory
```
Comment (bernardnormier) [inline python/modules/IcePy/ValueFactoryManager.h:27]: New name for "FactoryWrapper".

File: php/src/Communicator.cpp [hunk 4]
```diff
@@ -2094,15 +2053,11 @@ IcePHP::ValueFactoryManager::find(string_view id) const noexcept
 
     CommunicatorInfoIPtr info = p->second;
 
-    Ice::ValueFactoryPtr factory;
-    if (id.empty())
+    ValueFactoryPtr factory = info->findFactory(id);
+    if (!factory && id.empty())
```
Comment (pepone) [inline php/src/Communicator.cpp:2057]: we are missing a `std::move` call in the factory capture below.

File: python/modules/IcePy/ValueFactoryManager.cpp [hunk 5]
```diff
@@ -59,45 +59,33 @@ IcePy::ValueFactoryManager::~ValueFactoryManager()
...
-
-Ice::ValueFactoryFunc
-IcePy::ValueFactoryManager::find(string_view id) const noexcept
-{
-    Ice::ValueFactoryPtr factory = findCore(id);
 
     if (factory)
     {
-        return [factory](string_view type) -> shared_ptr<Ice::Value> { return factory->create(type); };
+        return [factory](string_view id) { return factory->create(id); };
```
Comment (pepone) [inline python/modules/IcePy/ValueFactoryManager.cpp:88]:
```suggestion
        return [factory = std::move(factory)](string_view id) { return factory->create(id); };
```
?

File: python/modules/IcePy/ValueFactoryManager.cpp [hunk 6]
```diff
@@ -110,7 +98,16 @@ IcePy::ValueFactoryManager::add(PyObject* valueFactory, string_view id)
 {
     try
     {
-        add(make_shared<FactoryWrapper>(valueFactory), id);
+        CustomValueFactoryPtr f = make_shared<CustomValueFactory>(valueFactory);
```
Comment (pepone) [inline python/modules/IcePy/ValueFactoryManager.cpp:101]: Do we need to create the CustomValueFactory here, or can we do it when we insert?

File: ruby/src/IceRuby/ValueFactoryManager.cpp [hunk 7]
```diff
@@ -66,95 +66,72 @@ IceRuby::ValueFactoryManager::ValueFactoryManager::create()
...
-}
 
-void IceRuby::ValueFactoryManager::addValueFactory(VALUE f, string_view id){
-    ICE_RUBY_TRY{add(make_shared<FactoryWrapper>(f), id);
-}
-ICE_RUBY_CATCH
+        _customFactories.insert(CustomFactoryMap::value_type(string{id}, f));
+    }
+    ICE_RUBY_CATCH
+    // clang-format off
```
Comment (pepone) [inline ruby/src/IceRuby/ValueFactoryManager.cpp:123]: Shouldn't this be `on` to re-enable?

**Review by pepone — APPROVED** [review]
👍 much simpler

File: php/src/Communicator.cpp (hunk 4 above)
Comment (bernardnormier) [reply php/src/Communicator.cpp:2057]: Fixed

File: python/modules/IcePy/ValueFactoryManager.cpp (hunk 5 above)
Comment (bernardnormier) [reply python/modules/IcePy/ValueFactoryManager.cpp:88]: Yes, fixed.

File: python/modules/IcePy/ValueFactoryManager.cpp (hunk 6 above)
Comment (bernardnormier) [reply python/modules/IcePy/ValueFactoryManager.cpp:101]: Switched to try_emplace, which is even more correct.

File: ruby/src/IceRuby/ValueFactoryManager.cpp (hunk 7 above)
Comment (bernardnormier) [reply ruby/src/IceRuby/ValueFactoryManager.cpp:123]: Yes, fixed

File: php/src/Communicator.cpp [hunk 8]
```diff
@@ -2013,77 +1999,50 @@ IcePHP::CommunicatorInfoI::getCommunicator() const
...
     }
+
     return nullptr;
 }
 
 void
 IcePHP::CommunicatorInfoI::destroyFactories(void)
 {
-    for (FactoryMap::iterator p = _factories.begin(); p != _factories.end(); ++p)
+    for (const auto& p : _customFactories)
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply php/src/Communicator.cpp]: Fixed as suggested

**Pull request merged**